### PR TITLE
Implementing cache for FIS SDK

### PR DESCRIPTION
**Pull request by ankitaj224**

### CODE REVIEW

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    // Cache entry is in delete state before syncing with Firebase backend
+    DELETED
+  }
+
+  private static final String SHARED_PREFS_NAME = "FirebaseInstallationIdCache";
+
+  private static final String FIREBASE_INSTALLATION_ID_KEY = "Fid";
+  private static final String AUTH_TOKEN_KEY = "AuthToken";
+  private static final String REFRESH_TOKEN_KEY = "RefreshToken";
+  private static final String TOKEN_CREATION_TIME_IN_SECONDS = "TokenCreationTime";
```
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: _KEY?

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  private static final String EXPIRES_IN_SECONDS = "ExpiresIn";
```
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: _KEY ?

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import android.content.SharedPreferences;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.firebase.FirebaseApp;
+
+/**
+ * A layer that locally caches a few Firebase Installation attributes on top the Firebase
+ * Installation backend API.
+ */
+public class FirebaseInstallationIdCache {
```
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: As this class is not used by developers, we can use a more brief name, for example FiidCache, same for the cache entry.

Not a comment, on your call.

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java (hunk 1 above)
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: TokenCreationTimeEpochSecs

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java (hunk 2 above)
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: ExpiresInSecs

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  @NonNull
+  public abstract String getAuthToken();
+
+  @NonNull
+  public abstract String getRefreshToken();
+
+  public abstract long getExpiresIn();
+
+  public abstract long getTokenCreationTime();
```
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java]: getTokenCreationEpochInSecs

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  @NonNull
+  public abstract FirebaseInstallationIdCache.CacheStatus getCacheStatus();
+
+  @NonNull
+  public abstract String getAuthToken();
+
+  @NonNull
+  public abstract String getRefreshToken();
+
+  public abstract long getExpiresIn();
```
Comment (diwu-arete) [inline firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java]: getExpiresInSecs

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java (hunk 1 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: Done. Thanks.

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java (hunk 3 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: Done. Thanks.

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java (hunk 2 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: Done. Thanks.

Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: Done. Thanks.

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java (hunk 5 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java]: Done. Thanks.

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java (hunk 4 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCacheEntryValue.java]: Done. Thanks.

File: firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java (hunk 1 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/local/FirebaseInstallationIdCache.java]: Done. Thanks.

**Pull request merged**